README.md
---
output: github_document
---

<!-- rmarkdown v1 -->

<!-- README.md is generated from README.Rmd. Please edit that file -->



# multilevelMatching

[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/multilevelMatching)](https://cran.r-project.org/package=multilevelMatching)
[![lifecycle](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://www.tidyverse.org/lifecycle/#stable)
[![Travis-CI Build Status](https://travis-ci.org/shuyang1987/multilevelMatching.svg?branch=master)](https://travis-ci.org/shuyang1987/multilevelMatching)
[![AppveyorCI Build status](https://ci.appveyor.com/api/projects/status/eu7vlcbu2j854cdo?svg=true)](https://ci.appveyor.com/project/BarkleyBG/multilevelmatching-3hh85)
[![Coverage status](https://codecov.io/gh/shuyang1987/multilevelMatching/branch/master/graph/badge.svg)](https://codecov.io/github/shuyang1987/multilevelMatching?branch=master)

### Propensity Score Matching and Subclassification in Observational Studies with Multi-Level Treatments 

Installation:


```r
devtools::install_github("shuyang1987/multilevelMatching")
```


### Visit the [package website](https://shuyang1987.github.io/multilevelMatching/)


# Description


This package implements methods to estimate causal effects from observational studies when there are 2+ distinct levels of treatment (i.e., "multilevel treatment") using matching estimators, as introduced in [Yang et al. (2016) Biometrics](https://doi.org/10.1111/biom.12505). Matching on covariates, and matching or stratification on modeled propensity scores, are made available. These methods require matching on only a scalar function of generalized propensity scores. For more information, see the Extended Description below or the main paper:

  - Yang, S., Imbens G. W., Cui, Z., Faries, D. E., & Kadziola, Z. (2016) Propensity Score Matching and Subclassification in Observational Studies with Multi-Level Treatments. *Biometrics*, 72, 1055-1065. https://doi.org/10.1111/biom.12505
  
    
Visit the [package website](https://shuyang1987.github.io/multilevelMatching/)


# Estimators available

- Matching on raw covariates: via `multiMatch()` and `multilevelMatchX()`
- Matching on estimated propensity scores: via `multiMatch()` and `multilevelGPSMatch()`
  - using ordinal logistic regression
  - using multinomial logistic regression
     - This method also provides two types of variance estimates
  - using user-provided propensity score values
     - This method does not provide variance estimates
- Stratification on propensity scores: via `multilevelGPSStratification()`

# Tutorial

This is a brief tutorial; an extended tutorial is provided in the vignette for [version 1.0.0](https://github.com/shuyang1987/multilevelMatching/releases/).
We will use the dataset provided with this package


```r
library(multilevelMatching)
simulated_data <- multilevelMatching::simulated_data
knitr::kable(head(simulated_data), digits = 2)
```



| outcome| treatment| covar1| covar2| covar3| covar4| covar5| covar6|
|-------:|---------:|------:|------:|------:|------:|------:|------:|
|   -5.13|         1|  -0.87|   0.24|   0.23|  -2.89|   0.21|      0|
|   -3.03|         1|   0.27|  -0.35|  -0.40|  -2.21|   0.07|      1|
|    3.05|         1|   1.42|   1.32|  -1.20|   0.06|   1.26|      1|
|   -6.09|         1|  -1.39|  -1.07|   1.12|  -2.36|   0.00|      0|
|   -2.46|         1|  -1.15|   0.95|   0.51|  -2.68|   0.07|      0|
|   -0.32|         1|   0.27|   0.42|  -0.45|   2.43|   0.60|      0|

We restructure the dataframe slightly, and use identifying names for the observations:


```r
outcome <- simulated_data$outcome
treatment <- simulated_data$treatment
covar_matrix <- as.matrix(
  simulated_data[ ,names(simulated_data) %in% paste0("covar", 1:6)]
)
identifying_names <- paste0(
  rep(letters[1:25],each = 12), rep(letters[1:25], 12)
)
names(treatment) <- identifying_names
```

## Matching on covariates


```r
set.seed(123)
fit <- multiMatch(
  Y = outcome,
  W = treatment,
  X = covar_matrix,
  match_on = "covariates"
)

fit
#> -------------- Causal estimates ---------------
#>         Param Trt1 Trt2   Estimate  Variance
#> 1 EY(2)-EY(1)    1    2 0.07927361 0.1792186
#> 2 EY(3)-EY(1)    1    3 0.86264929 0.1634754
#> 3 EY(3)-EY(2)    2    3 0.78337567 0.3221616
#> --- Matching on 'covariates' with M=1, J=1 ---
```

## Matching on the Estimated Generalized Propensity Score (GPS)

Propensity scores can be estimated with either of the following options

  - `match_on="multinom"` for multinomial logistic regression from `nnet::multinom()`
  - `match_on="polr"` for ordinal logistic regression from `MASS::polr()`
  - Or, estimated propensity scores can be supplied via the `X` argument when `match_on="existing"`
  

```r
match_on <- "multinom"
# match_on <- "polr" 

set.seed(123)
fit2 <- multiMatch(
  Y = outcome,
  W = treatment,
  X = covar_matrix,
  match_on = match_on,
  trimming = FALSE
)

fit
#> -------------- Causal estimates ---------------
#>         Param Trt1 Trt2   Estimate  Variance
#> 1 EY(2)-EY(1)    1    2 0.07927361 0.1792186
#> 2 EY(3)-EY(1)    1    3 0.86264929 0.1634754
#> 3 EY(3)-EY(2)    2    3 0.78337567 0.3221616
#> --- Matching on 'covariates' with M=1, J=1 ---
```


Please see the vignette for an extended tutorial.

# Extended Description

## Matching with 3 or more levels of treatment

In setting with where 3 or more levels of treatment (i.e., multilevel treatment), our goal is to estimate pairwise average treatment effects from a common population using matching methods.

This goal can not be acheived by matching one treatment with another one at a time, since the pairwise matched samples may differ from the target population systematically, and thus they are not compatitable. One implication is that from this approach, it is possible that treatment A is better than treatment B, treatment B is better than treatment C, and treatment C is better than treatment A. 

We focus on estimating the average values of potential outcomes for each treatment level by matching methods, which facilitate estimation of pairwise average treatment effects for a common population.

The estimation methods include generalized propensity score (GPS) matching, GPS stratification, matching with the full set of covariates, matching with the full set of GPS vector. Note that GPS matching and GPS straticication only require matching on a scalar function when estimating the average value of the potential outcome at a particular treatment level, which reduces the matching dimension to one, regardless of the number of covariates and the number of treatment levels. 

In order to ensure sufficient overlap, [Crump et al. (2009)](https://doi.org/10.1093/biomet/asn055)'s trimming method can be extended to this setting as well. 


# News

See [the News site](https://shuyang1987.github.io/multilevelMatching/news/index.html) for the changelog.

#### A note on `multiMatch()`

The `multiMatch()` function may return slightly different estimates than the original 2 matching functions in certain circumstances. We attempt to ensure that the functions implement are identical methods up to perhaps random number generation. Please file an issue if you have any questions or concerns.


NEWS.md

# multilevelMatching 1.0.0

- Original CRAN release version by Shu Yang ([@shuyang1987](https://github.com/shuyang1987/)) and Brian G Barkley ([@BarkleyBG](https://github.com/BarkleyBG/)).
- Added `multiMatch()` function to carry out all types of matching. This effectively combines `multilevelGPSMatch()` and `multilevelMatchX()` into one function. Features include:
  - Better output: 
     - tidier estimates: see `estimateTau()`
     - plus more information from under-the-hood: see the `impute_mat` object for matrix off all imputed potential outcomes.
  - Now allows for one-to-many matches for the main matching procedure (and imputing potential outomces). The user may specify `M_matches >=1`.
  - Now allows for one-to-many matches to estimate the variance component. The user may specify `J_var_matches >=1`.
  - Allowed for more user-specified arguments (for fitting PS models) i.e. `model_options`
- Divergence: Using `multiMatch()` with `match_on='existing'` does not always  return the same results as using `multilevelGPSMatch()` for matching on the existing (user-specified) generalized propensity scores. 
- Added S3 methods for `print` and `summary` for the `multiMatch` class
- Added S3 methods for `print` and `summary` for the `multiMatch` class
- Users can apply the `estimateTrtModel()` function before using `multiMatch()` to verify that the model fitted in `multiMatch()` is the same as the user desires

# multilevelMatching 0.1.0

- Original package version, released to GitHub by Shu Yang (@shuyang1987). See [Release v0.1.0](https://github.com/shuyang1987/multilevelMatching/releases/tag/v0.1.0).


R/multiMatch.R
#' Matching Estimators for Multiple Treatments from Yang et al. (2016).
#'
#'
#' This function carries out matching on covariates or on propensity scores, and
#' is similar to the 'legacy' functions \code{\link{multilevelMatchX}} and
#' \code{\link{multilevelGPSMatch}}.
#'
#' @param Y A response vector (1 x n)
#' @param W A treatment vector (1 x n) with numerical values indicating
#'   treatment groups
#' @param X A covariate matrix (p x n) with no intercept. When
#'   match_on="existing", then X must be a vector (1 x n) of user-specified
#'   propensity scores.
#' @param M_matches Number of matches per unit for imputing potential outcomes,
#'   as in Abadie and Imbens (2006).
#' @param J_var_matches Number of matches when estimating \eqn{\sigma^2(X,W)} as
#'   in Abadie and Imbens (2006).
#' @param trimming an indicator of whether trimming the sample to ensure overlap
#' @param match_on User specifies "covariates" to match on raw covariates, or
#'   "existing" to match on user-supplied propensity score values, or "polr" or
#'   "multinom" to fit a propensity score model.
#' @param model_options A list of the options to pass to propensity model.
#'   Currently under development. Can only pass reference level to multinomial
#'   logistic regression.
#'
#' @return A list of output from \code{estimateTau}, including at most: \itemize{
#'
#'   \item \code{tauestimate}:  a vector of estimates for pairwise treatment
#'   effects
#'
#'   \item \code{varestimate}:  a vector of variance estimates for tauestimate,
#'   using Abadie and Imbens (2006)'s method
#'
#'   \item \code{varestimateAI2016}:  a vector of variance estimates for
#'   tauestimate, when matching on the generalized propensity score, using
#'   Abadie & Imbens (2016)'s method. This variance estimate takes into account
#'   of the uncertainty in estimating the GPS.
#'
#'   \item \code{analysis_idx}: a list containing the indices_kept (analyzed)
#'   and indices_dropped (trimmed) based on Crump et al. (2009)'s method.
#'
#'   }
#'
#' @seealso \code{\link{multilevelMatchX}}; \code{\link{multilevelGPSMatch}}
#'
#' @references Yang, S., Imbens G. W., Cui, Z., Faries, D. E., & Kadziola, Z.
#'   (2016) Propensity Score Matching and Subclassification in Observational
#'   Studies with Multi-Level Treatments. Biometrics, 72, 1055-1065.
#'   \url{https://doi.org/10.1111/biom.12505}
#'
#'   Abadie, A., & Imbens, G. W. (2006). Large sample properties of matching
#'   estimators for average treatment effects. econometrica, 74(1), 235-267.
#'   \url{https://doi.org/10.1111/j.1468-0262.2006.00655.x}
#'
#'   Abadie, A., & Imbens, G. W. (2016). Matching on the estimated propensity
#'   score. Econometrica, 84(2), 781-807.
#'   \url{https://doi.org/10.3982/ECTA11293}
#'
#'   Crump, R. K., Hotz, V. J., Imbens, G. W., & Mitnik, O. A. (2009). Dealing
#'   with limited overlap in estimation of average treatment effects.
#'   Biometrika, 96(1), 187-199. \url{https://doi.org/10.1093/biomet/asn055}
#'
#' @examples
#'   sim_data <- multilevelMatching::simulated_data
#'   Y <- sim_data$outcome
#'   W <- sim_data$treatment
#'   X <- as.matrix(sim_data[ ,-(1:2)])
#'   names(Y) <- paste0("ID", 1:length(Y))
#'
#'   trimming <- FALSE
#'   method <- c("covariates", "polr", "multinom")[2]
#'
#'   multiMatch(Y,W,X,trimming=trimming,match_on=method)
#'
#'
#' @export
multiMatch <- function(
  Y,W,X,
  trimming = NULL,
  match_on,
  model_options = list(reference_level = sort(W)[1]),
  M_matches=1,
  J_var_matches=1
){

  ###                                          ###
  ### Step 1: Prepare data and other arguments ###
  ###                                          ###

  ## Argument checks and defensive programming
  prepared_data <- prepareData(
    Y = Y, W = W, X = X,
    match_on = match_on,
    trimming = trimming,
    model_options = model_options,
    M_matches = M_matches,
    J_var_matches = J_var_matches
  )
  X_covars <- prepared_data$X ## Necessary for AI2016 variance


  ## Fit treatment model to estimate propensity scores
  trt_model <- do.call(estimateTrtModel, prepared_data)
  prop_score_ests <- trt_model$prop_score_ests

  prepared_data <- append(
    prepared_data,
    list(
      M_matches = M_matches,
      J_var_matches = J_var_matches
    )
  )

  if (match_on != "covariates"){
    ## Instead of covariates, using propensity scores
    prepared_data$X <- prop_score_ests
  }


  ###                                          ###
  ### Step 2: Execute all matching procedures  ###
  ###                                          ###


  ## Carry out all estimation for all treatment levels
  matching_estimates <- do.call(matchAllTreatments, prepared_data)


  ###                                          ###
  ### Step 3: Organize into tidy, clean output ###
  ###                                          ###

  ## Tidy the results a little
  estimate_contrasts_args <- append(prepared_data, matching_estimates)
  results_list <- do.call(estimateTau, estimate_contrasts_args)
  tau_dfm <- results_list$tau_dfm

  impute_mat_sorted <- matching_estimates$Yiw
  impute_mat <- impute_mat_sorted[prepared_data$sorted_to_orig,]


  if (match_on == "multinom") {

  ###                                          ###
  ### Step 3B: Calculate AI2016 variance ests  ###
  ###                                          ###

    ## Collect a lot of objects from other function outputs
    est_var_AI2016_args <- append(estimate_contrasts_args, trt_model)
    est_var_AI2016_args$tau_dfm <- tau_dfm
    est_var_AI2016_args$X_covars <- X_covars

    ## Calculate AI2016 variance estimates for multinomial logistic regression
    est_var_AI2016 <- do.call(calcSigSqAI2016, est_var_AI2016_args)

    ## Add the AI2016 variance estimates to the tidy results dataframe
    tau_dfm$VarianceAI2016 <- est_var_AI2016$est_sigsq_AI2016
  }


  tidy_output <- list(
    results = tau_dfm,
    analysis_idx = prepared_data$analysis_idx,
    mu = results_list$mu_dfm,
    impute_mat = impute_mat,
    estimate_args = estimate_contrasts_args,
    model = trt_model$prop_score_model,
    propensity_scores = prop_score_ests,
    impute_match_data = matching_estimates$impute_match_data,
    impute_mat_sorted = impute_mat_sorted
  )

  if (match_on == "multinom") {
    ## Additional information from the AI2016 variance estimation
    tidy_output$AI2016_args <- est_var_AI2016$AI2016_args
  }

  class(tidy_output) <- "multiMatch"
  tidy_output
}


# #' Calculate Variance as in Abadie & Imbens (2016) JASA
# #'
# #' This function calculates the estimated variance of matching estimator when
# #' matching on propensity scores that were estimated with multinomial logistic
# #' regression. This method takes into account the uncertainty from the treatment
# #' model. This function does no "heavy lifting" in that it only assembles pieces
# #' that have been estimated elsewhere (i.e., see \code{\link{estSigSq}} to see
# #' how the matching procedures are carried out).
# #'
# #' @inheritParams multiMatch
# #' @inheritParams matchAllTreatments
# #' @param X_covars The matrix of covariates (not the matrix of estimated GPS)
# #' @param prop_score_ests The matrix of propensity scores
# #' @param vcov_coeff The variance-covariance (inverse fisher information) matrix
# #'   from the multinomial logistic regression model
# #' @param tau_dfm The dataframe of estimates and estimated standard errors
# #'   output from \code{estimateTau}
# #' @param match_mat_AI2016 Information on matches from
# #'   \code{\link{matchAllTreatments}}
# #'
# #'
# #' @references Abadie, A., & Imbens, G. W. (2016). Matching on the estimated
# #'   propensity score. Econometrica, 84(2), 781-807.
# #'   \url{https://doi.org/10.3982/ECTA11293}
# #'
# #'
# #' @return A list with the updated \code{tau_dfm} including a column for
# #'   \code{VarianceAI2016}, and a list object \code{AI2016_args} with
# #'   potentially helpful extra information.
calcSigSqAI2016 <- function(
  W, X_covars, Y, N,
  num_trts, trt_levels,
  match_mat_AI2016,
  prop_score_ests,
  vcov_coeff,
  tau_dfm,
  ...
){
  est_sigsq_AI2016 <- rep(NA, NROW(tau_dfm))

  ncol_X <- dim(X_covars)[2]
  c_matrix <- matrix(0, nrow=N, ncol=(dim(X_covars)[2]+1)*(num_trts-1))
  c_vector <- matrix(0, nrow=num_trts, ncol=(ncol_X+1)*(num_trts-1))

  for ( trt_kk in 1:num_trts ) {
    col_name <- nameCols(trt_levels[trt_kk])
    Y11 <- matrix( Y[match_mat_AI2016[,c(col_name)]], ncol=2, byrow=FALSE)
    mY11 <- apply(Y11,1,mean)
    for ( trt_k in 1:(num_trts-1) ) {
      for ( trt_jj in 1:(ncol_X+1) ) {
        if ( trt_jj==1 ) {}
        if ( trt_jj>1 ) {
          X11 <- matrix( X_covars[match_mat_AI2016[,c(col_name)],(trt_jj-1)],
                         ncol=2, byrow=FALSE)
          mX11 <- apply(X11,1,mean)

          c1_X1Y_temp <- apply((X11-mX11)*(Y11-mY11),1,sum)
          if ( trt_kk == (trt_k+1) ) {
            mult_factor <- (1-prop_score_ests[,trt_k+1])
          } else
            if ( trt_kk != (trt_k+1) ) {
              mult_factor <- (-1)*prop_score_ests[,trt_k+1]
            }
          c1_X1Y <- c1_X1Y_temp*mult_factor
          c_matrix[,(ncol_X+1)*(trt_k-1)+trt_jj] <- c1_X1Y
        }
      }
    }
    c_vector[trt_kk,] <- apply(c_matrix,2,mean)
  }

  for ( trt_jj in 1:(num_trts-1) ) {
    for ( trt_k in (trt_jj+1):num_trts ) {
      result_row_num <-
        which( tau_dfm$Param == nameContrast(
          trt1 = trt_levels[trt_jj], trt2 = trt_levels[trt_k]))

      est_sigsq_AI2016[result_row_num]  <-
        tau_dfm$Variance[result_row_num] -
        (
          t(c_vector[trt_jj,] + c_vector[trt_k,]) %*%
            vcov_coeff %*%
            ( c_vector[trt_jj,] + c_vector[trt_k,])
        )
    }
  }


  AI2016_args <- list(
    c_vector = c_vector,
    vcov_coeff = vcov_coeff,
    c_matrix = c_matrix
  )

  list(est_sigsq_AI2016 = est_sigsq_AI2016, AI2016_args = AI2016_args)
}
